[Bug]: Fuzzed AMMs hit 'panic: decimal division by zero'

### Problem encountered

Running market-sim fuzz testing against #11316 we hit division by zero panic.

Market-sim dump including logs and replay file:
[vega-sim-chwjwm78.zip](https://github.com/vegaprotocol/vega/files/15468506/vega-sim-chwjwm78.zip)

```
panic: decimal division by 0

goroutine 18382 [running]:
github.com/shopspring/decimal.Decimal.QuoRem({0x140181a3b00, 0xffffffd3}, {0x1400e346198, 0x0}, 0x10)
	/Users/caed/go/pkg/mod/github.com/vegaprotocol/decimal@v1.3.1-uint256/decimal.go:590 +0x244
github.com/shopspring/decimal.Decimal.DivRound({0x140181a3b00?, 0xfffffff0?}, {0x1400e346198?, 0xffffffd3?}, 0x10)
	/Users/caed/go/pkg/mod/github.com/vegaprotocol/decimal@v1.3.1-uint256/decimal.go:632 +0x3c
github.com/shopspring/decimal.Decimal.Div(...)
	/Users/caed/go/pkg/mod/github.com/vegaprotocol/decimal@v1.3.1-uint256/decimal.go:577
code.vegaprotocol.io/vega/core/execution/amm.(*Pool).priceForVolumeAtPosition(0x14015531c40?, 0x0, 0x1, 0x1?, 0x0?)
	/Users/caed/Projects/vega-market-sim/extern/vega/core/execution/amm/pool.go:668 +0x168
code.vegaprotocol.io/vega/core/execution/amm.(*Pool).OrderbookShape.func1(0x14015531c40, 0x0?, 0x0?)
	/Users/caed/Projects/vega-market-sim/extern/vega/core/execution/amm/pool.go:620 +0x4cc
code.vegaprotocol.io/vega/core/execution/amm.(*Pool).OrderbookShape(0x14015558780, 0x14016b69de0, 0x14016b69e20, 0x0)
	/Users/caed/Projects/vega-market-sim/extern/vega/core/execution/amm/pool.go:636 +0x5f0
code.vegaprotocol.io/vega/core/execution/common.(*MarketLiquidity).updateAMMCommitment(0x1400e6d8f00, 0xd)
	/Users/caed/Projects/vega-market-sim/extern/vega/core/execution/common/liquidity_provision_fees.go:155 +0x4dc
code.vegaprotocol.io/vega/core/execution/common.(*MarketLiquidity).OnTick(0x1400e6d8f00, {0x106830278, 0x14016b64300}, {0x0?, 0xedbec1560?, 0x0?})
	/Users/caed/Projects/vega-market-sim/extern/vega/core/execution/common/liquidity_provision.go:379 +0x190
code.vegaprotocol.io/vega/core/execution/future.(*Market).OnTick(0x1400e6de000, {0x106830278, 0x14016b64300}, {0x40?, 0x0?, 0x0?})
	/Users/caed/Projects/vega-market-sim/extern/vega/core/execution/future/market.go:1066 +0x554
code.vegaprotocol.io/vega/core/execution.(*Engine).OnTick(0x14000072300, {0x106830278, 0x14016b64300}, {0x1?, 0x10684f990?, 0x0?})
	/Users/caed/Projects/vega-market-sim/extern/vega/core/execution/engine.go:1186 +0x380
code.vegaprotocol.io/vega/core/vegatime.(*Svc).notify(...)
	/Users/caed/Projects/vega-market-sim/extern/vega/core/vegatime/service.go:107
code.vegaprotocol.io/vega/core/vegatime.(*Svc).SetTimeNow(0x14000d18080, {0x106830278, 0x14016b64300}, {0x40?, 0x108edfd20?, 0x0?})
	/Users/caed/Projects/vega-market-sim/extern/vega/core/vegatime/service.go:80 +0x244
code.vegaprotocol.io/vega/core/processor.(*App).OnBeginBlock(0x14000d85180, 0x4dd, {0x14014abd440, 0x40}, {0x14014abd440?, 0x0?, 0x0?}, {0x0, 0x0}, {0x108edfd20?, ...})
	/Users/caed/Projects/vega-market-sim/extern/vega/core/processor/abci.go:1170 +0x740
code.vegaprotocol.io/vega/core/blockchain/abci.(*App).FinalizeBlock(0x140009c4a80, {0x1065e0500?, 0x1?}, 0x140121c0480)
	/Users/caed/Projects/vega-market-sim/extern/vega/core/blockchain/abci/abci.go:167 +0x328
code.vegaprotocol.io/vega/cmd/vega/commands/node.(*appW).FinalizeBlock(0x140125a3a88?, {0x106830240?, 0x108edfd20?}, 0x14000b24000?)
	/Users/caed/Projects/vega-market-sim/extern/vega/cmd/vega/commands/node/app_wrapper.go:91 +0x30
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).processBlock(0x14000cb2000)
	/Users/caed/Projects/vega-market-sim/extern/vega/core/blockchain/nullchain/nullchain.go:198 +0x3e8
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).ForwardTime(0x14000cb2000, 0x3f5476a00)
	/Users/caed/Projects/vega-market-sim/extern/vega/core/blockchain/nullchain/nullchain.go:275 +0x194
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).handleForwardTime.func1()
	/Users/caed/Projects/vega-market-sim/extern/vega/core/blockchain/nullchain/server.go:132 +0x2c
created by code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).handleForwardTime in goroutine 18316
	/Users/caed/Projects/vega-market-sim/extern/vega/core/blockchain/nullchain/server.go:131 +0x1e0
 ```

### Observed behaviour

panic

### Expected behaviour

no panic

### Steps to reproduce

```bash
replay
```


### Software version

58d92cc52a281d7057ade283485de3dedf458f83

### Failing test

vega-market-sim fuzz test

### Jenkins run

_No response_

### Configuration used

```shell
{
  "genesis_time": "2023-05-09T10:00:00.000000000Z",
  "chain_id": "CUSTOM",
  "initial_height": "0",
  "consensus_params": {
    "block": {
      "max_bytes": "22020096",
      "max_gas": "-1",
      "time_iota_ms": "1"
    },
    "evidence": {
      "max_age_num_blocks": "100000",
      "max_age_duration": "172800000000000",
      "max_bytes": "1048576"
    },
    "validator": {
      "pub_key_types": ["ed25519"]
    },
    "version": {}
  },
  "validators": [
    {
      "address": "7DFD3744A773C166B36F170E4F5EF3E998703F4E",
      "pub_key": {
        "type": "tendermint/PubKeyEd25519",
        "value": "2lPO6I0p5uCBZzeoA2+ECJ2+57LtLVMjXpxH5za9TvE="
      },
      "power": "10",
      "name": ""
    }
  ],
  "app_hash": "",
  "app_state": {
    "assets": {
      "0000000000000000000000000000000000000000000000000000000000000000": {
        "name": "VOTE",
        "symbol": "VOTE",
        "decimals": 5,
        "quantum": "1",
        "min_lp_stake": "1",
        "source": {
          "builtin_asset": {
            "max_faucet_amount_mint": "100000000000000000"
          }
        }
      }
    },
    "validators": {
      "6NemqXSfN4zfZvAQeIANflEFqQ98izX9C0KLI2dITSY=": {
        "id": "c2b0ab69fb1bd19862c04df401920ea62d3e0bce41a50148240f42017824d17a",
        "vega_pub_key": "dfcf0fa8d67cfa11b3a511e320f3c47807a5e0735cc06dd9c2bc00f7ccc67117",
        "vega_pub_key_index": 1,
        "ethereum_address": "0x9e5BEEC6E56B28cCbd02864840B0f1e0125e42Ce",
        "tm_pub_key": "6NemqXSfN4zfZvAQeIANflEFqQ98izX9C0KLI2dITSY=",
        "info_url": "",
        "country": "",
        "name": "",
        "avatar_url": ""
      }
    },
    "network": {
      "replay_attack_threshold": 150
    },
    "network_parameters": {
      "blockchains.ethereumConfig": "{\"network_id\": \"3\", \"chain_id\": \"3\", \"collateral_bridge_contract\": { \"address\": \"0xa6F1E140daC13002Dfd9789D6dBA59117c717D7a\" }, \"confirmations\": 50, \"staking_bridge_contract\": { \"address\": \"0xfce2CC92203A266a9C8e67461ae5067c78f67235\", \"deployment_block_height\": 11001702}, \"multisig_control_contract\": {\"address\": \"0xCF6d41235911184fe6F35D47207813bFF3B91601\", \"deployment_block_height\": 12710009 } }",
      "blockchains.evmBridgeConfigs": "{\"configs\":[{\"network_id\": \"3\", \"chain_id\": \"3\", \"collateral_bridge_contract\": { \"address\": \"0xa6F1E140daC13002Dfd9789D6dBA59117c717D7a\" }, \"confirmations\": 50, \"multisig_control_contract\": {\"address\": \"0xCF6d41235911184fe6F35D47207813bFF3B91601\", \"deployment_block_height\": 12710009 } }]}",
      "blockchains.ethereumRpcAndEvmCompatDataSourcesConfig": "{\"configs\": []}",
      "governance.proposal.asset.maxClose": "8760h0m0s",
      "governance.proposal.asset.maxEnact": "8760h0m0s",
      "governance.proposal.asset.minClose": "1s",
      "governance.proposal.asset.minEnact": "1s",
      "governance.proposal.asset.minProposerBalance": "1",
      "governance.proposal.asset.minVoterBalance": "1",
      "governance.proposal.asset.requiredMajority": "0.66",
      "governance.proposal.asset.requiredParticipation": "0.00001",
      "governance.proposal.freeform.maxClose": "8760h0m0s",
      "governance.proposal.freeform.minClose": "1s",
      "governance.proposal.freeform.minProposerBalance": "1",
      "governance.proposal.freeform.minVoterBalance": "1",
      "governance.proposal.freeform.requiredMajority": "0.66",
      "governance.proposal.freeform.requiredParticipation": "0.00001",
      "governance.proposal.market.maxClose": "8760h0m0s",
      "governance.proposal.market.maxEnact": "8760h0m0s",
      "governance.proposal.market.minClose": "1s",
      "governance.proposal.market.minEnact": "2s",
      "governance.proposal.market.minProposerBalance": "1",
      "governance.proposal.market.minVoterBalance": "1",
      "governance.proposal.market.requiredMajority": "0.66",
      "governance.proposal.market.requiredParticipation": "0.00001",
      "governance.proposal.updateAsset.maxClose": "720h",
      "governance.proposal.updateAsset.maxEnact": "720h",
      "governance.proposal.updateAsset.minClose": "1s",
      "governance.proposal.updateAsset.minEnact": "1s",
      "governance.proposal.updateAsset.minProposerBalance": "1",
      "governance.proposal.updateAsset.minVoterBalance": "1",
      "governance.proposal.updateAsset.requiredMajority": "0.66",
      "governance.proposal.updateAsset.requiredParticipation": "0.09",
      "governance.proposal.updateMarket.maxClose": "8760h0m0s",
      "governance.proposal.updateMarket.maxEnact": "8760h0m0s",
      "governance.proposal.updateMarket.minClose": "1s",
      "governance.proposal.updateMarket.minEnact": "1s",
      "governance.proposal.updateMarket.minProposerBalance": "1",
      "governance.proposal.updateMarket.minProposerEquityLikeShare": "0",
      "governance.proposal.updateMarket.minVoterBalance": "1",
      "governance.proposal.updateMarket.requiredMajority": "0.66",
      "governance.proposal.updateMarket.requiredMajorityLP": "0.66",
      "governance.proposal.updateMarket.requiredParticipation": "0.00001",
      "governance.proposal.updateMarket.requiredParticipationLP": "0",
      "governance.proposal.updateNetParam.maxClose": "8760h0m0s",
      "governance.proposal.updateNetParam.maxEnact": "8760h0m0s",
      "governance.proposal.updateNetParam.minClose": "1s",
      "governance.proposal.updateNetParam.minEnact": "1s",
      "governance.proposal.updateNetParam.minProposerBalance": "1",
      "governance.proposal.updateNetParam.minVoterBalance": "1",
      "governance.proposal.updateNetParam.requiredMajority": "0.5",
      "governance.proposal.updateNetParam.requiredParticipation": "0.00001",
      "governance.proposal.referralProgram.maxClose": "8760h0m0s",
      "governance.proposal.referralProgram.maxEnact": "8760h0m0s",
      "governance.proposal.referralProgram.minClose": "1s",
      "governance.proposal.referralProgram.minEnact": "1s",
      "governance.proposal.referralProgram.minProposerBalance": "1",
      "governance.proposal.referralProgram.minVoterBalance": "1",
      "governance.proposal.referralProgram.requiredMajority": "0.5",
      "governance.proposal.referralProgram.requiredParticipation": "0.00001",
      "governance.proposal.VolumeDiscountProgram.maxClose": "8760h0m0s",
      "governance.proposal.VolumeDiscountProgram.maxEnact": "8760h0m0s",
      "governance.proposal.VolumeDiscountProgram.minClose": "1s",
      "governance.proposal.VolumeDiscountProgram.minEnact": "1s",
      "governance.proposal.VolumeDiscountProgram.minProposerBalance": "1",
      "governance.proposal.VolumeDiscountProgram.minVoterBalance": "1",
      "governance.proposal.VolumeDiscountProgram.requiredMajority": "0.5",
      "governance.proposal.VolumeDiscountProgram.requiredParticipation": "0.00001",
      "governance.proposal.transfer.maxClose": "8760h0m0s",
      "governance.proposal.transfer.maxEnact": "8760h0m0s",
      "governance.proposal.transfer.minClose": "1s",
      "governance.proposal.transfer.minEnact": "1s",
      "governance.proposal.transfer.minProposerBalance": "1",
      "governance.proposal.transfer.minVoterBalance": "1",
      "governance.proposal.transfer.requiredMajority": "0.5",
      "governance.proposal.transfer.requiredParticipation": "0.00001",
      "limits.markets.maxPeggedOrders": "1500",
      "limits.markets.proposePerpetualEnabled": "1",
      "limits.markets.ammPoolEnabled": "1",
      "market.auction.maximumDuration": "168h",
      "market.auction.minimumDuration": "1s",
      "market.fee.factors.infrastructureFee": "0.0005",
      "market.fee.factors.makerFee": "0.0002",
      "market.liquidity.bondPenaltyParameter": "0.1",
      "market.liquidity.earlyExitPenalty": "0.1",
      "market.liquidity.maximumLiquidityFeeFactorLevel": "0.03",
      "market.liquidity.minimum.probabilityOfTrading.lpOrders": "1e-6",
      "market.liquidity.probabilityOfTrading.tau.scaling": "10.0",
      "market.liquidity.providersFeeCalculationTimeStep": "1s",
      "market.liquidity.sla.nonPerformanceBondPenaltyMax": "0.5",
      "market.liquidity.sla.nonPerformanceBondPenaltySlope": "2",
      "market.liquidity.stakeToCcyVolume": "1.0",
      "market.liquidity.targetstake.triggering.ratio": "0.25",
      "market.liquidityProvision.minLpStakeQuantumMultiple": "5000",
      "market.liquidityProvision.shapes.maxSize": "100",
      "market.margin.scalingFactors": "{\"search_level\": 1.1, \"initial_margin\": 1.5, \"collateral_release\": 1.7}",
      "market.monitor.price.defaultParameters": "{\"triggers\": [{\"auction_extension\": 300, \"horizon\": 43200, \"probability\": \"0.9999999\"}] }",
      "market.stake.target.scalingFactor": "0.0001",
      "market.stake.target.timeWindow": "1h",
      "market.value.windowLength": "2h0m0s",
      "network.checkpoint.timeElapsedBetweenCheckpoints": "5m",
      "network.floatingPointUpdates.delay": "5m",
      "network.markPriceUpdateMaximumFrequency": "5s",
      "network.transaction.defaultgas": "1",
      "network.transactions.maxgasperblock": "3000000",
      "network.transactions.minBlockCapacity": "32",
      "network.validators.ersatz.multipleOfTendermintValidators": "0",
      "network.validators.ersatz.rewardFactor": "0.85",
      "network.validators.incumbentBonus": "0.05",
      "network.validators.minimumEthereumEventsForNewValidator": "3",
      "network.validators.multisig.numberOfSigners": "13",
      "network.validators.tendermint.number": "13",
      "reward.asset": "0000000000000000000000000000000000000000000000000000000000000000",
      "reward.staking.delegation.competitionLevel": "1.3",
      "reward.staking.delegation.delegatorShare": "0.8",
      "reward.staking.delegation.maxPayoutPerEpoch": "9890000000000000000000",
      "reward.staking.delegation.maxPayoutPerParticipant": "700000000000000000000",
      "reward.staking.delegation.minimumValidatorStake": "3000000000000000000000",
      "reward.staking.delegation.minValidators": "5",
      "reward.staking.delegation.optimalStakeMultiplier": "3.0",
      "reward.staking.delegation.payoutDelay": "0h",
      "reward.staking.delegation.payoutFraction": "1.0",
      "rewards.marketCreationQuantumMultiple": "1",
      "snapshot.interval.length": "300",
      "spam.pow.difficulty": "15",
      "spam.pow.hashFunction": "sha3_24_rounds",
      "spam.pow.increaseDifficulty": "0",
      "spam.pow.numberOfPastBlocks": "100",
      "spam.pow.numberOfTxPerBlock": "1000",
      "spam.protection.delegation.min.tokens": "100000000000000000",
      "spam.protection.max.batchSize": "30",
      "spam.protection.max.delegations": "360",
      "spam.protection.max.proposals": "300",
      "spam.protection.max.votes": "300",
      "spam.protection.maxUserTransfersPerEpoch": "10000",
      "spam.protection.minimumWithdrawalQuantumMultiple": "10",
      "spam.protection.minMultisigUpdates": "100000000000000000000",
      "spam.protection.proposal.min.tokens": "2000000000000000000000",
      "spam.protection.voting.min.tokens": "1000000000000000000",
      "spam.protection.applyReferral.min.funds": "0",
      "transfer.fee.factor": "0.001",
      "transfer.minTransferQuantumMultiple": "0",
      "transfer.fee.maxQuantumAmount": "100",
      "transfer.feeDiscountMinimumTrackedAmount": "0.001",
      "transfer.feeDiscountDecayFraction": "0.5",
      "validator.performance.scaling.factor": "0",
      "validators.delegation.minAmount": "1",
      "validators.epoch.length": "2m",
      "validators.vote.required": "0.67",
      "referralProgram.maxReferralTiers": "10",
      "referralProgram.maxReferralRewardFactor": "1",
      "referralProgram.maxReferralDiscountFactor": "1",
      "referralProgram.maxPartyNotionalVolumeByQuantumPerEpoch": "10000",
      "referralProgram.minStakedVegaTokens": "0",
      "referralProgram.maxReferralRewardProportion": "1",
      "market.amm.minCommitmentQuantum": "1"
    },
    "network_limits": {
      "propose_market_enabled": true,
      "propose_asset_enabled": true,
      "bootstrap_block_count": 0
    },
    "checkpoint": {
      "load_hash": ""
    }
  }
}
```


### Relevant log output

```shell
panic: decimal division by 0

goroutine 18382 [running]:
github.com/shopspring/decimal.Decimal.QuoRem({0x140181a3b00, 0xffffffd3}, {0x1400e346198, 0x0}, 0x10)
	/Users/caed/go/pkg/mod/github.com/vegaprotocol/decimal@v1.3.1-uint256/decimal.go:590 +0x244
github.com/shopspring/decimal.Decimal.DivRound({0x140181a3b00?, 0xfffffff0?}, {0x1400e346198?, 0xffffffd3?}, 0x10)
	/Users/caed/go/pkg/mod/github.com/vegaprotocol/decimal@v1.3.1-uint256/decimal.go:632 +0x3c
github.com/shopspring/decimal.Decimal.Div(...)
	/Users/caed/go/pkg/mod/github.com/vegaprotocol/decimal@v1.3.1-uint256/decimal.go:577
code.vegaprotocol.io/vega/core/execution/amm.(*Pool).priceForVolumeAtPosition(0x14015531c40?, 0x0, 0x1, 0x1?, 0x0?)
	/Users/caed/Projects/vega-market-sim/extern/vega/core/execution/amm/pool.go:668 +0x168
code.vegaprotocol.io/vega/core/execution/amm.(*Pool).OrderbookShape.func1(0x14015531c40, 0x0?, 0x0?)
	/Users/caed/Projects/vega-market-sim/extern/vega/core/execution/amm/pool.go:620 +0x4cc
code.vegaprotocol.io/vega/core/execution/amm.(*Pool).OrderbookShape(0x14015558780, 0x14016b69de0, 0x14016b69e20, 0x0)
	/Users/caed/Projects/vega-market-sim/extern/vega/core/execution/amm/pool.go:636 +0x5f0
code.vegaprotocol.io/vega/core/execution/common.(*MarketLiquidity).updateAMMCommitment(0x1400e6d8f00, 0xd)
	/Users/caed/Projects/vega-market-sim/extern/vega/core/execution/common/liquidity_provision_fees.go:155 +0x4dc
code.vegaprotocol.io/vega/core/execution/common.(*MarketLiquidity).OnTick(0x1400e6d8f00, {0x106830278, 0x14016b64300}, {0x0?, 0xedbec1560?, 0x0?})
	/Users/caed/Projects/vega-market-sim/extern/vega/core/execution/common/liquidity_provision.go:379 +0x190
code.vegaprotocol.io/vega/core/execution/future.(*Market).OnTick(0x1400e6de000, {0x106830278, 0x14016b64300}, {0x40?, 0x0?, 0x0?})
	/Users/caed/Projects/vega-market-sim/extern/vega/core/execution/future/market.go:1066 +0x554
code.vegaprotocol.io/vega/core/execution.(*Engine).OnTick(0x14000072300, {0x106830278, 0x14016b64300}, {0x1?, 0x10684f990?, 0x0?})
	/Users/caed/Projects/vega-market-sim/extern/vega/core/execution/engine.go:1186 +0x380
code.vegaprotocol.io/vega/core/vegatime.(*Svc).notify(...)
	/Users/caed/Projects/vega-market-sim/extern/vega/core/vegatime/service.go:107
code.vegaprotocol.io/vega/core/vegatime.(*Svc).SetTimeNow(0x14000d18080, {0x106830278, 0x14016b64300}, {0x40?, 0x108edfd20?, 0x0?})
	/Users/caed/Projects/vega-market-sim/extern/vega/core/vegatime/service.go:80 +0x244
code.vegaprotocol.io/vega/core/processor.(*App).OnBeginBlock(0x14000d85180, 0x4dd, {0x14014abd440, 0x40}, {0x14014abd440?, 0x0?, 0x0?}, {0x0, 0x0}, {0x108edfd20?, ...})
	/Users/caed/Projects/vega-market-sim/extern/vega/core/processor/abci.go:1170 +0x740
code.vegaprotocol.io/vega/core/blockchain/abci.(*App).FinalizeBlock(0x140009c4a80, {0x1065e0500?, 0x1?}, 0x140121c0480)
	/Users/caed/Projects/vega-market-sim/extern/vega/core/blockchain/abci/abci.go:167 +0x328
code.vegaprotocol.io/vega/cmd/vega/commands/node.(*appW).FinalizeBlock(0x140125a3a88?, {0x106830240?, 0x108edfd20?}, 0x14000b24000?)
	/Users/caed/Projects/vega-market-sim/extern/vega/cmd/vega/commands/node/app_wrapper.go:91 +0x30
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).processBlock(0x14000cb2000)
	/Users/caed/Projects/vega-market-sim/extern/vega/core/blockchain/nullchain/nullchain.go:198 +0x3e8
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).ForwardTime(0x14000cb2000, 0x3f5476a00)
	/Users/caed/Projects/vega-market-sim/extern/vega/core/blockchain/nullchain/nullchain.go:275 +0x194
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).handleForwardTime.func1()
	/Users/caed/Projects/vega-market-sim/extern/vega/core/blockchain/nullchain/server.go:132 +0x2c
created by code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).handleForwardTime in goroutine 18316
	/Users/caed/Projects/vega-market-sim/extern/vega/core/blockchain/nullchain/server.go:131 +0x1e0
```

--- a/core/execution/amm/pool.go
+++ b/core/execution/amm/pool.go
@@ -598,33 +598,42 @@ func (p *Pool) OrderbookShape(from, to *num.Uint, idgen *idgeneration.IDGenerato
 				// we are in "approximation" mode with a step bigger than a tick and have stepped over the AMM's
 				// fair-price. We need to split this step into two, a buy order from current -> fp, and a sell
 				// from fp -> next
-				volume := uint64(num.DeltaV(position, 0))
-				price := p.priceForVolumeAtPosition(volume, types.OtherSide(side), 0, fairPrice)
-				buys = append(buys, p.makeOrder(volume, price, side, idgen))
+
+				fairPosition := p.getPosition()
+				volume := uint64(num.DeltaV(position, fairPosition))
+
+				// our approximate bigger steps may be less than a price level away from the fair-price in which case
+				// the volume will be < 1 and we can just ignore an order for this segment since we are already *roughly* expanding into orders.
+				if volume != 0 {
+					price := p.priceForVolumeAtPosition(volume, types.OtherSide(side), fairPosition, fairPrice)
+					buys = append(buys, p.makeOrder(volume, price, side, idgen))
+				}
 
 				// we've step through fair-price now so orders will becomes sells
 				side = types.SideSell
 				current = fairPrice
-				position = 0
+				position = fairPosition
 			}
 
 			nextPosition := cu.positionAtPrice(p.sqrt, num.Min(next, cu.high))
 			volume := uint64(num.DeltaV(position, nextPosition))
 
-			if side == types.SideBuy {
-				price := current
-				if approx {
-					price = p.priceForVolumeAtPosition(volume, types.OtherSide(side), nextPosition, next)
+			if volume != 0 {
+				if side == types.SideBuy {
+					price := current
+					if approx {
+						price = p.priceForVolumeAtPosition(volume, types.OtherSide(side), nextPosition, next)
+					}
+					order := p.makeOrder(volume, price, side, idgen)
+					buys = append(buys, order)
+				} else {
+					price := next
+					if approx {
+						price = p.priceForVolumeAtPosition(volume, types.OtherSide(side), position, current)
+					}
+					order := p.makeOrder(volume, price, side, idgen)
+					sells = append(sells, order)
 				}
-				order := p.makeOrder(volume, price, side, idgen)
-				buys = append(buys, order)
-			} else {
-				price := next
-				if approx {
-					price = p.priceForVolumeAtPosition(volume, types.OtherSide(side), position, current)
-				}
-				order := p.makeOrder(volume, price, side, idgen)
-				sells = append(sells, order)
 			}
 
 			// if we're calculating buys and we hit fair price, switch to sells
@@ -655,6 +664,10 @@ func (p *Pool) PriceForVolume(volume uint64, side types.Side) *num.Uint {
 // priceForVolumeAtPosition returns the price the AMM is willing to trade at to match with the given volume if its position and fair-price
 // are as given.
 func (p *Pool) priceForVolumeAtPosition(volume uint64, side types.Side, pos int64, fp *num.Uint) *num.Uint {
+	if volume == 0 {
+		panic("cannot calculate price for zero volume trade")
+	}
+
 	x, y := p.virtualBalances(pos, fp, side)
 
 	// dy = x*y / (x - dx) - y

--- a/core/execution/amm/pool_test.go
+++ b/core/execution/amm/pool_test.go
@@ -482,7 +482,7 @@ func testOrderbookShapeStepOverFairPrice(t *testing.T) {
 	ensurePosition(t, p.pos, position, num.UintZero())
 	require.Equal(t, "25", p.pool.BestPrice(nil).String())
 
-	ensurePosition(t, p.pos, position, num.UintZero())
+	ensurePositionN(t, p.pos, position, num.UintZero(), 2)
 	buys, sells := p.pool.OrderbookShape(low, base, nil)
 	assert.Equal(t, 3, len(buys))
 	assert.Equal(t, 8, len(sells))
@@ -492,7 +492,7 @@ func testOrderbookShapeStepOverFairPrice(t *testing.T) {
 	assert.Equal(t, 0, len(buys))
 	assert.Equal(t, 11, len(sells))
 
-	ensurePosition(t, p.pos, position, num.UintZero())
+	ensurePositionN(t, p.pos, position, num.UintZero(), 2)
 	buys, sells = p.pool.OrderbookShape(low, high, nil)
 	assert.Equal(t, 2, len(buys))
 	assert.Equal(t, 9, len(sells))

